File: day008_string_traversal/related_digit_string/solution.java
import java.util.Scanner;

class Solution {

    // Optimal: O(N) Time - Single Pass
    public static void solve_digit_string() {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String S = scanner.nextLine();
        
        if (S.isEmpty()) {
            System.out.println("NO");
            return;
        }
        
        if (S.matches("\\d+")) { // Java regex is often cleanest
            System.out.println("YES");
        } else {
            System.out.println("NO");
        }
        scanner.close();
    }

    // Brute Force: O(N) Time - Manual Character.isDigit() check
    public static void solve_digit_string_brute() {
        Scanner scanner = new Scanner(System.in);
        if (!scanner.hasNextLine()) return;
        String S = scanner.nextLine();
        
        if (S.isEmpty()) {
            System.out.println("NO");
            return;
        }
        
        for (int i = 0; i < S.length(); i++) {
            if (!Character.isDigit(S.charAt(i))) {
                System.out.println("NO");
                scanner.close();
                return;
            }
        }
        System.out.println("YES");
        scanner.close();
    }

    public static void main(String[] args) {
        solve_digit_string();
    }
}

File: day008_string_traversal/related_digit_string/solution.cpp
#include <iostream>
#include <string>
#include <cctype> // for isdigit

using namespace std;

// Optimal: O(N) Time - Single Pass
void solve_digit_string() {
    string S;
    if (!getline(cin, S)) return;

    if (S.empty()) {
        cout << "NO" << endl; // Empty string is not a digit string
        return;
    }

    for (char c : S) {
        if (!isdigit(c)) {
            cout << "NO" << endl;
            return;
        }
    }
    
    cout << "YES" << endl;
}

// Brute Force: O(N) Time - Manual check using char limits
void solve_digit_string_brute() {
    string S;
    if (!getline(cin, S)) return;
    
    if (S.empty()) {
        cout << "NO" << endl;
        return;
    }
    
    for (char c : S) {
        if (c < '0' || c > '9') {
            cout << "NO" << endl;
            return;
        }
    }
    
    cout << "YES" << endl;
}

int main() {
    solve_digit_string();
    return 0;
}